### PR TITLE
Process procs now properly utilize deltatime when implementing rates, timers and probabilities

### PR DESCRIPTION
**Pull request by Donkie**
## About The Pull Request
Most of our codebase define rates and probabilities "per tick". Repaired health per tick, chance of breaking per tick, etc etc.

This is quite dumb since it means if we want to change the tickrate of things, we need to re-tune every single thing that uses that tickrate. We also have some code that is impacted by multiple tickrates which create even more complex dynamics. Changing the tickrate of systems should impact performance, stability and accuracy of the solution, not impact the state trajectory.

Another reason for this is that physical equations often require a time component. The volume pump has incorrectly been spitting 400L/s for years since the atmos tick is at 0.5s, yet the process proc for the pump makes no effort in treating this behaviour.

Some details about my implementation:
- My original plan was to send a "world.time - last_run" float as a parameter to the process proc, which would yield even more accurate physics simulations, and will also circumvent the issues that can arise when subsystems are ignored due to too much tick usage. However I decided to go with the #define macro route so that as many of these calculations will happen at compile time instead of runtime, to make sure that this PR doesn't impact performance.
- Most `prob` in process procs have been replaced with my homemade `DT_PROB` macro. It's explained in code, but it correctly calculates the adjusted probability based on different tickrates. The downside to this macro is that it requires a power operator so it can be expensive if done runtime, however most of the situations contain only hardcoded numbers and they are thus calculated at compiletime.
- Removed the `seconds` or `wait * 0.1` arguments that is sent in the process proc because literally nothing uses it.

If you think I should rename the constants to something else, or move their definitions somewhere else, please say so.

If this is appreciated, I will also deltatimefy the other subsystems like mobs.

## Why It's Good For The Game
Easier for people to tweak subsystem's tickrates in order to optimize server performance.

## Changelog
:cl:
code: Process procs now properly utilize deltatime when implementing rates, timers and probabilities, which paves way for easier subsystem tickrate adjustments.
/:cl:

### CODE REVIEW

**Comment by Qustinnus** [comment]
I've been waiting for something like this for quite some time. It'll probably go unnoticed but this'll make balancing way more easier if people can think in units per second instead of units per tick.

One big thing I noticed from looking at the code is that you didn't actually convert any of the numbers from tick to second, which means that everything that for example anything that runs in SSMOBS will now run twice as quick (As it ticks once every two seconds, but you now compensate for deltatime rather than just using tick). This'll need to be fixed by manually dividing all of the numbers by these static delta-times.

**Comment by Qustinnus** [comment]
also tick defines should probably go in the _Defines/ file that has subsystem stuff

**Comment by Donkie** [comment]
I actually have converted all numbers. Some DTs however are at 1 sec so they didn't need touching.

Alright I will move them

File: code/modules/mob/living/silicon/pai/pai.dm
Comment (Qustinnus) [inline code/modules/mob/living/silicon/pai/pai.dm:317]: not this one

File: code/modules/antagonists/changeling/powers/mutations.dm
Comment (Qustinnus) [inline code/modules/antagonists/changeling/powers/mutations.dm:499]: why would this one need 2 DTs? it affects mobs but it runs on the objects process() so you only need SSOBJ_DT

Comment (Donkie) [reply code/modules/antagonists/changeling/powers/mutations.dm:499]: Since it uses reagents_metabolism constant I assume it's supposed to fill just enough for the mob to metabolize. Thus both DTs are needed.

**Comment by Qustinnus** [comment]
my bad yeah I misread some code. this looks good then but i'm not sure about the need of using 2 deltatimes on one process(), can you explain why thats necesary?

File: code/modules/antagonists/changeling/powers/mutations.dm
Comment (Qustinnus) [reply code/modules/antagonists/changeling/powers/mutations.dm:499]: right, makes sense.

File: code/game/objects/effects/anomalies.dm
Comment (Donkie) [inline code/game/objects/effects/anomalies.dm:68]: Make macro const instead if this isn't changed anywhere

File: code/game/objects/items/RSF.dm
Comment (Donkie) [inline code/game/objects/items/RSF.dm:196]: max_matter needs changing

File: code/modules/clothing/spacesuits/hardsuit.dm
Comment (Donkie) [inline code/modules/clothing/spacesuits/hardsuit.dm:86]: Make a macro for this instead and use that for the other Geiger counter as well

File: code/modules/ninja/suit/n_suit_verbs/energy_net_nets.dm
Comment (Donkie) [inline code/modules/ninja/suit/n_suit_verbs/energy_net_nets.dm:48]: Confirm this bad boy

**Comment by optimumtact** [comment]
@MrStonedOne

**Comment by Qustinnus** [comment]
if this is merged I'll PR some guidelines for it in contributing.MD

**Comment by optimumtact** [comment]
Most of our codebase define rates and probabilities "per tick". Repaired health per tick, chance of breaking per tick, etc etc.

This is quite dumb since it means if we want to change the tickrate of things, we need to re-tune every single thing that uses that tickrate.

Don't call it per tick, call it per cycle, that way people won't be confused between it and the world time ticking

**Comment by MrStonedOne** [comment]
The entire system seems to be based on seconds, but all of byond and all our other systems are based on dediseconds. I foresee this causing issues and confusion, and I don't know that there is any benefit to doing it this way and creating two parallel units of times. 

@optimumtact what do you think about this? I feel like we should standardize on one or other and avoid systems that expect something different (outside of tick counts as a unit, for the rare thing that makes sense to scale based on game ticks). IIRC, we have a few cases of things that are per-second in code, and i remember many cases in old #coderbus of this confusing people.

The defines bit I'm also not sure about, for one, admins can change the fire rate of a system, I do often when debugging lag on the live servers. I get there is a cpu/speed component to the decision, but how much of a concern is that really? Such a system would work better and be easier for programmers to use if it just took a subsystem as an argument, and looked at that subsystem's actual wait value.   
(or even the delta between the last run and the current run (if you do this, you'll need to change master.dm:535 to use `queue_node.queued_time` instead of `world.time` for last_fire, and use `SS.queued_time-SS.last_fire` when calculating the delta, otherwise you'll compare the *start* of a subsystem's fire to when it last *ended*, skewing results on subsystems that take more then a game tick to run)).

> now properly utilize deltatime

This implies deltatime was a thing before, (and nothing was properly using it), but this is not true is it?

Also it seems like there should be some macros for math-operation-over-time. ie `SS_MATH(+, num, per-ds-change, SSblah)` (but less lame)

One last tip: byond will compile away any constant math operation inside of a (), and macro arguments that are constant count. it will try do this in other situations, but you can use `()` to be sure.

**Comment by Qustinnus** [comment]
> This implies deltatime was a thing before, (and nothing was properly using it), but this is not true is it?


some process() procs had static deltatime as arg

**Comment by optimumtact** [comment]
however, if we use deciseconds we align with world.time, and that might be less confusing

File: code/modules/clothing/spacesuits/hardsuit.dm
Comment (MrStonedOne) [reply code/modules/clothing/spacesuits/hardsuit.dm:86]: the old code is basically just `MC_AVERAGE()`

The new code? fuck if i know.

Given the intention of this, I don't think you need to scale the math here, but if you did, the multiplication form would be more readable.

`rad_record = (rad_count * (0.20 * (SSOBJ_DT/2))) + (rad_record * (0.80 * (SSOBJ_DT/2)))`

Comment (Donkie) [reply code/modules/clothing/spacesuits/hardsuit.dm:86]: It's a low pass filter, which is basically a moving average filter. I realise that this new form looks awful so I will make some macro that makes it more readable.

**Comment by Donkie** [comment]
I had not considered the fact that you might want to tweak the subsystem wait on the fly. That could definitely be something that's worth embracing because it let's us test if things still work fine for longer tickrates. 

Deltatime is a concept that always should exist when dealing with game loops, because that is how you properly implement rate of change. And as @Qustinnus  said, someone made some feeble attempt to let everyone use Deltatime when calling `process()` but it seems like nobody took notice. 

Good point about the queued_time.

I'm not quite sure what you mean with the suggested SS_MATH macro
@MrStonedOne

If the defines were to use deciseconds it would just require me to shove ` SECONDS` after every case of using them, increasing bloat. Why not use seconds as a unit to begin with then?

File: code/modules/ninja/suit/n_suit_verbs/energy_net_nets.dm
Comment (Donkie) [reply code/modules/ninja/suit/n_suit_verbs/energy_net_nets.dm:48]: Mental note: /obj/structure/energy_net is not being processed at all

**Comment by Donkie** [comment]
Did some more changes. @MrStonedOne if you believe it's better to have a dynamic queued_time deltatime instead of this constant one system, just say and I'll go ahead and change it up. Most of the grunt work has been done already.

**Comment by Donkie** [comment]
Sorry to bug you @MrStonedOne but what do you think?

**Comment by optimumtact** [comment]
get this mergeable fyi, if he doesn't reply within a couple of days we'll just progress with this.

**Comment by Donkie** [comment]
@optimumtact hmu when you want to merge this if it isn't merge-ready

File: code/__DEFINES/subsystems.dm
Comment (MrStonedOne) [inline code/__DEFINES/subsystems.dm:251]: change all of this to use the wait. 

```dm
#define SSAIR_DT (SSair.wait/10)
...
```

I don't like duplicating info in a way that can create a desync, or making it so nobody can change a wait without changing the same value in multiple places, and we do still need to support var edited wait times.

I might also suggest adding SECONDS to the define, to avoid confusion (such as `SSAIR_DT_SECONDS`), but don't feel like blocking on that.

Comment (Donkie) [reply code/__DEFINES/subsystems.dm:251]: The waits have been updated to use these defines, so it's only defined in one place.

Comment (MrStonedOne) [reply code/__DEFINES/subsystems.dm:251]: Thats not what i asked.

The waits should remain dediseconds, both in given unit and stored unit.

Changing them to use these defines doesn't help with varedited waits. 

Change these defines to refer to the subsystem wait, keep the subsystem wait defined in the subsystem's file.

Comment (Donkie) [reply code/__DEFINES/subsystems.dm:251]: Oh I see what you mean, then we will lose the benefit of having this available at compile time. If we're going that route I'd prefer if it was actually dynamic like we discussed with queued_time etc.

Comment (MrStonedOne) [reply code/__DEFINES/subsystems.dm:251]: it won't quite work as i expected, some procs and sleep and resume after fire() has ended, the calculation for those cases will be wildly off. my plan is to go through and use those defines to quickly make everything use it.

and for the reference, subsystems will exist before any object code runs. (outside of stat()), so i don't forsee an issue to just making those defines refer to the wait by global var reference.

Comment (Donkie) [reply code/__DEFINES/subsystems.dm:251]: Alright, I'll convert the defines then

**Comment by Donkie** [comment]
@MrStonedOne implemented your suggestions

File: code/game/objects/effects/alien_acid.dm
Comment (AnturK) [inline code/game/objects/effects/alien_acid.dm:78]: Probably round given the switch below. (or changing that)

File: code/game/objects/items/robot/robot_upgrades.dm
Comment (AnturK) [inline code/game/objects/items/robot/robot_upgrades.dm:368]: Using these defines look pretty bad outside of simple defines to me. Not very helpful with understanding the meaning either since the result is not in seconds.

File: code/game/turfs/open/lava.dm
Comment (AnturK) [inline code/game/turfs/open/lava.dm:141]: That doesn't make much sense, whatever the armor was it was cut down to 50 before ?

**Review by AnturK — COMMENTED** [review]
Good PR, you should add a readme string to base process explaining the use of these defines.

File: code/game/turfs/open/lava.dm
Comment (Donkie) [reply code/game/turfs/open/lava.dm]: You're right, that's pretty weird.

File: code/game/objects/items/robot/robot_upgrades.dm
Comment (Donkie) [reply code/game/objects/items/robot/robot_upgrades.dm]: Alright, fair enough

File: code/game/objects/effects/alien_acid.dm
Comment (Donkie) [reply code/game/objects/effects/alien_acid.dm:80]: Done

File: code/game/machinery/slotmachine.dm
Comment (Donkie) [reply code/game/machinery/slotmachine.dm]: Done

**Review by MrsTonedOne — APPROVED** [review]

**Comment by Donkie** [comment]
@optimumtact merge pls

**Comment by AnturK** [comment]
Going to testmerge for a moment and if nothing breaks obviously i'll merge.

**Comment by AnturK** [comment]
Seems there's no major issues so good to go when conflicts are fixed.

**Comment by Donkie** [comment]
@AnturK Fixed conflicts

**Pull request merged**
